src/NLog/Internal/MruCache.cs

```diff
@@ -0,0 +1,185 @@
+                    return false;   // Already exists
+                }
+                else
+                {
+                    var version = ++_currentVersion;
+                    if (_dictionary.Count >= _maxCapacity)
+                    {
+                        // Make 3 sweeps:
+                        //  1) Versions below _currentVerison - _maxCapacity / 2
+                        //  2) Versions below _currentVerison - _maxCapacity / 10
+                        //  3) Versions below _currentVerison - 2
+                        var pruneKeys = new List<TKey>(_dictionary.Count);
+                        for (int i = 1; i <= 3; ++i)
+                        {
+                            long oldVersion = version - 2;
+                            switch (i)
+                            {
+                                case 1: oldVersion = version - (int)(_maxCapacity / 1.5); break;
+                                case 2: oldVersion = version - _maxCapacity / 10; break;
+                            }
+                            foreach (var element in _dictionary)
+                            {
+                                if (element.Value.Version <= oldVersion)
+                                {
+                                    pruneKeys.Add(element.Key);
+                                    if (_dictionary.Count - pruneKeys.Count < _maxCapacity / 1.5)
+                                    {
+                                        i = 3;
+                                        break;  // Do not clear the entire cache
+                                    }
+                                }
+                            }
+                        }
+
+                        foreach (var pruneKey in pruneKeys)
+                        {
+                            _dictionary.Remove(pruneKey);
+                        }
+
+                        if (_dictionary.Count >= _maxCapacity)
+                        {
+                            _dictionary.Clear();
+                        }
+                    }
+
+                    _dictionary.Add(key, new MruItem(value, version));
+                    return true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Lookup existing item in cache
+        /// </summary>
+        /// <param name="key">Dictionary Key</param>
+        /// <param name="value">Dictionary Value [out]</param>
+        /// <returns>Key found</returns>
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            MruItem item;
+            try
+            {
+                if (!_dictionary.TryGetValue(key, out item))
+                {
+                    value = default(TValue);
+                    return false;
+                }
+            }
+            catch
+            {
+                item = default(MruItem);    // Too many people in the same room
+            }
+
+            if (item.Version != _currentVersion)
```

Comment:
Please some docs here, thanks! (I guess  retrieving the sane is here on purpose)